<!DOCTYPE html>
<html>
  <head>
    <meta charset="utf-8">
    <meta name="generator" content="pandoc">
    <title>TIB2017: NGS – from fastq to variant annotation</title>
    <link rel="shortcut icon" type="image/x-icon" href="/favicon.ico" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
    <link rel="stylesheet" type="text/css" href="css/bootstrap/bootstrap.css" />
    <link rel="stylesheet" type="text/css" href="css/bootstrap/bootstrap-theme.css" />
    <link rel="stylesheet" type="text/css" href="css/swc.css" />
    <link rel="alternate" type="application/rss+xml" title="PBI - Transcriptomica" href="http://bi.ira.cinvestav.mx:8282/"/>
    <meta charset="UTF-8" />
    <!-- HTML5 shim, for IE6-8 support of HTML5 elements -->
    <!--[if lt IE 9]>
      <script src="http://html5shim.googlecode.com/svn/trunk/html5.js"></script>
    <![endif]-->
  </head>
  <body class="lesson">
    <div class="container card">
      <div class="banner">
        <a href="https://mxbiobankworkshop.wordpress.com/" title="Practical workshop on Large-Scale
            Genomic Data Analyses: GWAS in structured populations">
          <img alt="MX Biobank Logo" src="img/mxbiobank_logo_reduced.png" />
        </a>
      </div>
      <article>
      <div class="row">
        <div class="col-md-10 col-md-offset-1">
                    <a href="index.html"><h1 class="title">NGS – from fastq to variant annotation</h1></a>
          <h2 class="subtitle">Read alignment</h2>
          <section class="objectives panel panel-warning">
<div class="panel-heading">
<h2 id="learning-objectives"><span class="glyphicon glyphicon-certificate"></span>Learning Objectives</h2>
</div>
<div class="panel-body">
<ul>
<li>Align sequencing data to a reference genome</li>
<li>Understand how to interpret high-throughput sequencing alignment data</li>
<li>First approach to the SAM and BAM coordinate formats</li>
</ul>
</div>
</section>
<p>We will use the fastq files that we used in the previous practice, as well as the reference genome of our organism, <em>Saccharomyces pombe</em>:</p>
<div class="sourceCode"><pre class="sourceCode bash"><code class="sourceCode bash">$ <span class="kw">cd</span> .. 
$ <span class="kw">pwd</span> </code></pre></div>
<pre class="output"><code>/Data/VariantCalling</code></pre>
<p>Characteristics of the experiment:</p>
<ul>
<li>Yeast genome: 12.5 Mbp; 16 chromosomes</li>
<li>Whole genome sequencing</li>
<li>Paired-end reads, 108bp, one library, 2 lanes</li>
</ul>
<h2 id="mapping-the-reads-to-the-genome">Mapping the reads to the genome</h2>
<p>Once we verify that the reads are in the correct format, we will align the reads and transcripts to the genome using BWA.</p>
<p>You can find the manual in the following<a href="http://bio-bwa.sourceforge.net/bwa.shtml">link</a>.</p>
<div class="sourceCode"><pre class="sourceCode bash"><code class="sourceCode bash">$ <span class="kw">bwa</span></code></pre></div>
<pre class="output"><code>Program: bwa (alignment via Burrows-Wheeler transformation)
Version: 0.7.17-r1188
Contact: Heng Li &lt;lh3@sanger.ac.uk&gt;

Usage:   bwa &lt;command&gt; [options]

Command: index         index sequences in the FASTA format
         mem           BWA-MEM algorithm
         fastmap       identify super-maximal exact matches
         pemerge       merge overlapping paired ends (EXPERIMENTAL)
         aln           gapped/ungapped alignment
         samse         generate alignment (single ended)
         sampe         generate alignment (paired ended)
         bwasw         BWA-SW for long queries

         shm           manage indices in shared memory
         fa2pac        convert FASTA to PAC format
         pac2bwt       generate BWT from PAC
         pac2bwtgen    alternative algorithm for generating BWT
         bwtupdate     update .bwt to the new format
         bwt2sa        generate SA from BWT and Occ

Note: To use BWA, you need to first index the genome with `bwa index&#39;.
      There are three alignment algorithms in BWA: `mem&#39;, `bwasw&#39;, and
      `aln/samse/sampe&#39;. If you are not sure which to use, try `bwa mem&#39;
      first. Please `man ./bwa.1&#39; for the manual.</code></pre>
<p>First we will generate a bwa index for the genome:</p>
<div class="sourceCode"><pre class="sourceCode bash"><code class="sourceCode bash">$ <span class="kw">bwa</span> index -a is Saccharomyces_cerevisiae.EF4.68.dna.toplevel.fa</code></pre></div>
<pre class="output"><code>[bwa_index] Pack FASTA... 0.12 sec
[bwa_index] Construct BWT for the packed sequence...
[bwa_index] 7.56 seconds elapse.
[bwa_index] Update BWT... 0.07 sec
[bwa_index] Pack forward-only FASTA... 0.07 sec
[bwa_index] Construct SA from BWT and Occ... 3.67 sec
[main] Version: 0.7.17-r1188
[main] CMD: bwa index -a is Saccharomyces_cerevisiae.EF4.68.dna.toplevel.fa
[main] Real time: 12.089 sec; CPU: 11.490 sec</code></pre>
<div class="sourceCode"><pre class="sourceCode bash"><code class="sourceCode bash">$ <span class="kw">ls</span> Saccharomyces*</code></pre></div>
<pre class="output"><code>Saccharomyces_cerevisiae.EF4.68.dna.toplevel.fa  
Saccharomyces_cerevisiae.EF4.68.dna.toplevel.fa.amb
Saccharomyces_cerevisiae.EF4.68.dna.toplevel.fa.ann
Saccharomyces_cerevisiae.EF4.68.dna.toplevel.fa.bwt
Saccharomyces_cerevisiae.EF4.68.dna.toplevel.fa.pac
Saccharomyces_cerevisiae.EF4.68.dna.toplevel.fa.sa</code></pre>
<p>All new files that do not end with .fa form the bwa index.</p>
<p>For aligning the reads we will be using bwa mem, we check its manual:</p>
<div class="sourceCode"><pre class="sourceCode bash"><code class="sourceCode bash">$ <span class="kw">bwa</span> mem</code></pre></div>
<p>Some commonly used options include:</p>
<p>-t number of threads/processors to use – see PBS script at end of workbook -p Assume the first input query file is interleaved paired-end FASTA/Q. See the command description for details. -a Output all found alignments for single-end or unpaired paired-end reads. These alignments will be flagged as secondary alignments.</p>
<p>We will use default options:</p>
<div class="sourceCode"><pre class="sourceCode bash"><code class="sourceCode bash">$ <span class="kw">bwa</span> mem Saccharomyces_cerevisiae.EF4.68.dna.toplevel.fa lane1/s-7-1.fastq lane1/s-7-2.fastq <span class="kw">&gt;</span> lane1.sam</code></pre></div>
<p>We omit the output.</p>
<p>We explore the result, which is a file in SAM format.</p>
<div class="sourceCode"><pre class="sourceCode bash"><code class="sourceCode bash">$ <span class="kw">head</span> lane1.sam</code></pre></div>
<pre class="output"><code>@SQ SN:I    LN:230218
@SQ SN:II   LN:813184
@SQ SN:III  LN:316620
@SQ SN:IV   LN:1531933
@SQ SN:IX   LN:439888
@SQ SN:Mito LN:85779
@SQ SN:V    LN:576874
@SQ SN:VI   LN:270161
@SQ SN:VII  LN:1090940
@SQ SN:VIII LN:562643</code></pre>
<div class="sourceCode"><pre class="sourceCode bash"><code class="sourceCode bash">$ <span class="kw">tail</span> lane1.sam</code></pre></div>
<pre class="output"><code>IL29_4505:7:120:19585:18832#2   77  *   0   0   *   *   0   0   GATCGGAAGAGCGGTTCAGCAGGAAATGCCGAGACCGATCTCCGATGTTTATCTCGTATGCCGTCTTCTGCTTGAAAAAAAAACAAACCACATAACTACATCTCCACG    CCBCBDCCCBBBBBBCBCBBBDCBCBAACBCBBBBAAAAAB=BB&lt;?BABAAA.ABBAA?89BBCCBA;AA-B0B@AAAAB(--(%$&amp;$-3$%$&#39;*6&lt;).635%&#39;+*/*    AS:i:0  XS:i:0
IL29_4505:7:120:19585:18832#2   141 *   0   0   *   *   0   0   GATCGGAAGAGCGGCGGGGAGGGGAAGAGGGGAGATCTCGGGGGGCGCCGGATCATTAAAAAAAAAAAAAAAAAAAGGACGTAAACCGTGATATCTCCGTCCGCGTGA    @64&lt;;?27134)2)0.&amp;1&#39;$8=9.14-&#39;2&amp;1(/*/,,2424*49&#39;&amp;1%0,&amp;)/&#39;-*0(43&gt;&gt;&gt;&gt;&gt;&gt;&gt;&gt;33&amp;.*&amp;)%$&#39;&amp;$&amp;$(%$$$$*$%)%&amp;)$%($),%$+$&amp;$&amp;    AS:i:0  XS:i:0
IL29_4505:7:120:19710:18114#2   77  *   0   0   *   *   0   0   GATCGGAAGAGCGGTTCAGCAGGGAATGCCGAGACCGATCTCCGATGTTTATCTCGTATGCCGTCTTCTGCTTGAAAAAAAACACACAAACACATGACTGTGACGATG    AABBBBBBBABBBBABABB:BBBAABBBBBBBBAB?AABABBABBABAABAA5?ABA?BB@&lt;B@@&lt;9&gt;&gt;;)=:?4AA=9B,3(0$(-&#39;(%-)0&amp;%*.,&#39;*$$,&amp;&amp;-13    AS:i:0  XS:i:0
IL29_4505:7:120:19710:18114#2   141 *   0   0   *   *   0   0   GATCGGAAGAGCGTCGGGTAGGGAAAGAGGGAAGATCTCGGGGGCCGCCGTATCATTAAAAAAAAAAAAAACACAACAATCAAACTCTACCACACCTGGCACACGCGA    ?4+=@?3333,&amp;7%-:,/0((;&lt;-6/).5*2&amp;16)(02*77&amp;52&amp;&amp;1&#39;*+*-//0333-5=71144%)+$$&#39;%$%&amp;&amp;&amp;(%%%-%*%-$$*&amp;&amp;*$$$$*$-$,%&#39;%%&#39;%    AS:i:0  XS:i:0
IL29_4505:7:120:19764:17570#2   77  *   0   0   *   *   0   0   GATCGGAAGAGCGGTTCAGCAGGAATGCCGAGACCGATCTCCGAAGTTTATCTCGTATGCCGTCTTCTGCTTGAAAAAAAAACTAAGACACGATACCATCTCACCGCT    BBCB@?AAA?BABAA@C:CAAAB&lt;AC?@7C&lt;B?8BACABBB?BB&lt;C?@?B?*9?==B?@??=ABAAABA?BAAAA0;97;(.%+&lt;1*&amp;*&amp;(%&amp;&amp;$(&#39;,&amp;(%+.,+5=%    AS:i:0  XS:i:0
IL29_4505:7:120:19764:17570#2   141 *   0   0   *   *   0   0   GATCGGAAGCGCGGCGTGTGGGGCCAGAGTGTAGAGCTGGGTGGTCGCCGCATACTTTAAAAAAGATAATGAGATCATGCGAAGTGTATCTGCAAGTGGATCTAATAC    &lt;@+&gt;=9(93);72&#39;=9,,)+222&amp;&#39;(5*,.7&#39;10&amp;&#39;3,&#39;6*&#39;1&lt;%&amp;9)*4%/&amp;(%0.%*)5&lt;&gt;4%/$%%%)$(&#39;(+,%.*-&amp;)/(%&#39;(&amp;%(*5&amp;)&#39;$&amp;()31*,).-%    AS:i:0  XS:i:0
IL29_4505:7:120:19773:11846#2   77  *   0   0   *   *   0   0   CCGATGTTTATCTCGTATGCCGTCTTCTGCTTGAAAACACAAAAGAAACATACTCTATCAAGATACATGCTAGTAACGAAAACACACACATGACTCCACACTGTGATG    A@@*@@6@@@@&gt;@@&lt;:@;@@:A;A?@/@4@@6@@@:A+;4&gt;&amp;9,$$,&#39;%%&#39;1-(%6%/:,(*/%&amp;&amp;%&#39;(7,55(04*&#39;&#39;84%8,-.%+1/(&amp;)&amp;&amp;)%&amp;%&amp;%%&amp;*$%1.    AS:i:0  XS:i:0
IL29_4505:7:120:19773:11846#2   141 *   0   0   *   *   0   0   GATTTCATAAACACAGGCCTGAATGTAGACAGGTCACTAGAGATACCGTCACAAACAAAACATCAATACTAACAACAACTACTACAATCACCACACACATCACCACAC    *.(=4&#39;0*&#39;1&#39;/;6/-(**(20((46&#39;(&#39;2**(5-*,5(**&#39;8764.*.((&lt;/&#39;./&#39;&amp;3(.6,*(;-&#39;*-*(8,+&#39;&#39;*/(&lt;0*(*&#39;.&#39;((-7&#39;*3/&#39;/)&#39;(7)&#39;5/&#39;0    AS:i:0  XS:i:0
IL29_4505:7:120:19817:15837#2   77  *   0   0   *   *   0   0   ATTTTTAGTATACGTGAATAATGTATTTTTTAAATTGGTTATTGCGTAGGAACACAATATTTTAAACCTTATCGATTAATTTTACTTGACTGAGTACGAGCGCCAGCT    ABA8A@BB8BBBBBA?BBBB;ABAB@&gt;9?B/;@8B@&gt;A??B6:BA@AAB?BB/=A:?@;AB@BB8ABB@BA@B&gt;/ABB&gt;BB-?B/A3.&lt;=.=&gt;3&lt;B/&lt;B82&#39;077.*9    AS:i:0  XS:i:0
IL29_4505:7:120:19817:15837#2   141 *   0   0   *   *   0   0   AAAAAAAAAAAAAACGGTGGACTAAAACACCCGGTCGTTGAAACAAAATTTGATAAATGTCTAAAGGCATTTACATGTCCTACATCCAATCATATGACTTTTGTTTCC    =2=,=,(-8179+4(&#39;(-))1&amp;)-&#39;(,/2**&#39;&#39;&amp;)*83(&amp;:21))7)&#39;6++%1)(&#39;3%&#39;))%%&#39;2%&amp;(:2&#39;&#39;1&amp;*2&amp;&#39;&#39;&amp;&#39;8+040&amp;3*2,&lt;&#39;2-%3&amp;&amp;&#39;--&#39;(*&#39;%-    AS:i:0  XS:i:0</code></pre>
<h3 id="the-sam-format">The SAM format</h3>
<p>Let’s see a smaller example of this format. Suppose we have the following alignment:</p>
<pre class="output"><code>Coor    12345678901234 5678901234567890123456789012345
ref AGCATGTTAGATAA**GATAGCTGTGCTAGTAGGCAGTCAGCGCCAT
+r001/1       TTAGATAAAGGATA*CTG
+r002        aaaAGATAA*GGATA
+r003      gcctaAGCTAA
+r004                    ATAGCT..............TCAGC
-r003                           ttagctTAGGC
-r001/2                                       CAGCGGCAT</code></pre>
<p>The corresponding SAM format will be the following:</p>
<pre class="output"><code>@HD VN:1.5  SO:coordinate
@SQ SN:ref  LN:45
r001    99  ref 7   30  8M2I4M1D3M  =   37  39  TTAGATAAAGGATACTG   *
r002    0   ref 9   30  3S6M1P1I4M  *   0   0   AAAAGATAAGGATA  *
r003    0   ref 9   30  5S6M    *   0   0   GCCTAAGCTAA *   SA:Z:ref,29,-,6H5M,17,0;
r004    0   ref 16  30  6M14N5M *   0   0   ATAGCTTCAGC *
r003    2064    ref 29  17  6H5M    *   0   0   TAGGC   *   SA:Z:ref,9,+,5S6M,30,1;
r001    147 ref 37  30  9M  =   7   -39 CAGCGGCAT   *   NM:i:1</code></pre>
<p>The SAM format is a plain text format that allows us to save sequencing data in ASCII format delimited by tabs.</p>
<p>It is made up of two core sections:</p>
<ul>
<li>The headers</li>
<li>The alignment</li>
</ul>
<p>The section of the <strong>header</strong> starts with the character <code>@</code> followed by one of the codes of two letters that denote the characteristics of the alignments in this file. Each line is delimited by tabs and, in addition to the lines that begin with <code>@CO</code>, each data field has the format<code>TAG:VALUE</code>, where <code>TAG</code> is a string Two characters that define the format and content of <code>VALUE</code>.</p>
<p>The header is not indispensable but contains information about the version of the file as well as if it is ordered or not. Therefore it is advisable to include it.</p>
<p>The <strong>alignment section</strong> contains the following information:</p>
<ol style="list-style-type: decimal">
<li><strong>QNAME</strong> Name of the reference, QNAME (SAM) / Name of the read (BAM). It is used to group alignments that are together, as in the case of alignments of paired end reads or a read that appears in multiple alignments.</li>
<li><strong>FLAG</strong> Information set describing the alignment. Provides the following information:
<ul>
<li>Are there multiple fragments?</li>
<li>Are all the fragments well aligned?</li>
<li>Is this fragment aligned?</li>
<li>Has the following fragment not been aligned?</li>
<li>Is the reference the reverse string?</li>
<li>Is the following fragment the reverse chain?</li>
<li>Is this the first fragment?</li>
<li>Is this the last fragment?</li>
<li>Is this a secondary alignment?</li>
<li>Did this reading fail quality filters?</li>
<li>Is this reading a PCR or optical duplicate?</li>
</ul></li>
<li><strong>RNAME</strong> Name of the reference sequence.</li>
<li><strong>POS</strong> Left alignment position (base 1).</li>
<li><strong>MAP</strong> Alignment quality.</li>
<li><strong>CIGAR</strong> CIGAR chain.</li>
<li><strong>RNEXT</strong> Name of the paired end (mate) or the next read.</li>
<li><strong>PNEXT</strong> Position of the pair (mate) or the next read.</li>
<li>** TLEN** Length of the alignment.</li>
<li><strong>SEQ</strong> The test sequence of this alignment (in this case the sequence of the read).</li>
<li><strong>QUAL</strong> The read quality.</li>
<li><strong>TAGs</strong> Additional information.</li>
</ol>
<aside class="callout panel panel-info">
<div class="panel-heading">
<h2 id="cadenas-cigar"><span class="glyphicon glyphicon-pushpin"></span>Cadenas CIGAR</h2>
</div>
<div class="panel-body">
<p>The sequence aligned to the reference may have additional bases that are not in the reference or the read may be missing bases that are in the reference. The CIGAR string is a string that encodes each base and the characteristics of each in alignment.</p>
<p>For example, the string CIGAR:</p>
<pre class="output"><code>CIGAR: 3M1I3M1D5M</code></pre>
<p>indicates that the first 3 bases of the reading aligns with the reference (3M), the next base does not exist in the reference (1I), the following 3 bases align with the reference (3M), the next base does not exist in the reading (1D), and 5 more bases align with the reference (5M).</p>
</div>
</aside>
<p>As you can see these files contain a lot of information that can be analyzed using scripts that show alignment statistics. Programs like <a href="http://broadinstitute.github.io/picard/">Picard</a> perform this type of analysis.</p>
<p>The compressed version of the SAM files is known as BAM (binary sam). Let’s convert the SAM file to BAM using samtools:</p>
<div class="sourceCode"><pre class="sourceCode bash"><code class="sourceCode bash">$ <span class="kw">samtools</span> view -b lane1.sam -o lane1.bam</code></pre></div>
<p>We do not open the BAM files because, since they are in binary format, they are illegible. However, we have to perform two last steps to visualize the results:</p>
<div class="sourceCode"><pre class="sourceCode bash"><code class="sourceCode bash">$ <span class="kw">samtools</span> sort lane1.bam -o lane1_sorted.bam
$ <span class="kw">samtools</span> index lane1_sorted.bam</code></pre></div>
<p>The first step orders the results by their coordinates and the second one creates indexes to speed up the display using a browser.</p>
        </div>
      </div>
      </article>
      <div class="footer">
        <a class="label swc-blue-bg" href="mailto:selene.fernandez.valverde@gmail.com">Contacto</a>
        <a class="label swc-blue-bg" href="LICENSE.html">Licencia</a>
      </div>
    </div>
    <!-- Javascript placed at the end of the document so the pages load faster -->
    <script src="http://software-carpentry.org/v5/js/jquery-1.9.1.min.js"></script>
    <script src="css/bootstrap/bootstrap-js/bootstrap.js"></script>
    <script src='https://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML'></script>
  </body>
</html>
